reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,383 @@
+            ComponentPlan newPlan = componentPlanSearch(root, nextPendingKeys, choices);
+
+            double newPlanCost = newPlan.cost(root);
+            if (bestPlan == null || newPlanCost < bestPlanCost) {
+                bestPlan = newPlan;
+                bestPlanCost = newPlanCost;
+            }
+
+            choices.remove(mode);
+        }
+        assert bestPlan != null;
+        return bestPlan;
+    }
+
+    private void initialiseOrderingDependencies(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> inputBounds) {
+        Set<Variable> currentBounds = new HashSet<>(inputBounds);
+        for (Resolvable<?> resolvable : ordering) {
+            Set<Variable> resolvableBounds = iterate(consideredVariables(resolvable)).filter(currentBounds::contains).toSet();
+            initialiseDependencies(conjunctionNode, resolvable, resolvableBounds);
+            if (!resolvable.isNegated()) currentBounds.addAll(consideredVariables(resolvable));
+        }
+    }
+
+    private void initialiseDependencies(ConjunctionNode conjunctionNode, Resolvable<?> resolvable, Set<Variable> resolvableBounds) {
+        if (resolvable.isConcludable()) {
+            Set<ResolvableConjunction> cyclicDependencies = conjunctionNode.cyclicDependencies(resolvable.asConcludable());
+            for (CallMode callMode : triggeredCalls(resolvable.asConcludable(), resolvableBounds, Optional.empty())) {
+                recursivelyGenerateOrderingChoices(callMode);
+                if (!cyclicDependencies.contains(callMode.conjunction)) {
+                    plan(callMode); // Acyclic dependencies can be fully planned
+                }
+            }
+        } else if (resolvable.isNegated()) {
+            iterate(resolvable.asNegated().disjunction().conjunctions()).forEachRemaining(conjunction -> {
+                CallMode callMode = new CallMode(conjunction, resolvableBounds);
+                recursivelyGenerateOrderingChoices(callMode);
+                plan(callMode);
+            });
+        }
+    }
+
+    private void recursivelyGenerateOrderingChoices(CallMode callMode) {
+        if (!candidateOrderings.containsKey(callMode)) {
+            candidateOrderings.put(callMode, null); // Guard
+            ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(callMode.conjunction);
+            answerCountEstimator.buildConjunctionModel(callMode.conjunction);
+
+            // Two orderings for the same CallMode with identical cyclic-concludable modes are interchangeable in the component-plan
+            //      -> We only need to keep the cheaper one.
+            Map<Set<Pair<Concludable, Set<Variable>>>, OrderingChoice> bestChoice = new HashMap<>();
+            PartialOrderReductionSearch porSearch = new PartialOrderReductionSearch(logicMgr.compile(callMode.conjunction), callMode.bounds);
+            for (List<Resolvable<?>> ordering : porSearch.allOrderings()) {
```

Comment:
i'd still call this variable `plans`, even though the search returns orderings. This is where the domain changes: the user of an API is allowed to bring it into a new domain (a plan is a specialisation of an order), so we can name it so here.

Reply:
We've been somewhat consistent that a plan has a cost associated with it.

Reply:
yeah alright